mllib/src/main/scala/org/apache/spark/ml/param/shared/sharedParams.scala

```diff
@@ -357,4 +357,17 @@ private[ml] trait HasWeightCol extends Params {
   /** @group getParam */
   final def getWeightCol: String = $(weightCol)
 }
+
+/**
+ * Trait for the solver algorithm for optimization.
+ */
+private[ml] trait HasSolver extends Params {
+  /**
+   * Param for solver algorithm. If this is not set or empty, we select the solver algorithm automatically.
+   */
+  final val solver: Param[String] = new Param[String](this, "solver", "the solver algorithm for optimization. If this is not set or empty, default value is 'auto'.")
```

Comment:
Shall we have a validation function here to check the ```solver``` is set with legal value? Such as if users set ```solver``` with ```normal``` but the number of features is more than 4096, throwing exception here. [WeightedLeastSquares](https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/optim/WeightedLeastSquares.scala#L173) will check this after job starting, but I think we should detect this error earlier.